### DISCUSSION TITLE
Driving a piezo connected across 2 pins to increse the output volume

### DISCUSSION THREAD

**Discussion by adrianblakey** [micropython/micropython#12954] [Hardware & Peripherals] [answered]
In order to increase the volume of a piezo connected to a Pi Pico W, it's connected across 2 pins - 20, 21. So one pin has to be driven high while the other is low then swapped around for the 2nd half of the oscillation period.
How do I arrange the code to drive both pins with pwm?

This works - but it's normal volume:

```
        p20 = Pin(20, Pin.OUT)
        p21 = Pin(21, Pin.OUT)
        p20.high()
        p21.low()
        buzzerUp = PWM(Pin(21))
        buzzerDown = PWM(Pin(20))
        buzzerUp.duty_u16(2 ** 15) # setting this to 
        buzzerDown.duty_u16(0) # setting this to  2 ** 15 results in no sound
        buzzerUp.freq(500)
        buzzerDown.freq(500)
```

**Comment by davefes** [comment]
Is the piezo not polarity-sensitive?    Use the PWM output to drive a transistor, say off 5V 12V whatever the voltage rating of the piezo is.

**Reply by 2dof** [reply]
based on your description - I notice that You use solution more like to create polyphony solution not loudness control.

as @davefes suggested  piezo buzzer  is polarity sensitive. You should connect (-) terminal to the GND , and (+) terminal to the PWM output (via transistor driver). (see[ about piezo buzzer] (https://arduinogetstarted.com/tutorials/arduino-piezo-buzzer#content_about_piezo_buzzer) ).

If You want variable control loudness then you can do this by potentiometer ( as manual control) or based on OP-AMP build  a direct-current voltage control circuit for inputting changeable gain (see [Volume control circuit of piezoelectric buzzer ](https://patents.google.com/patent/CN102291646A/en)

**Reply by adrianblakey** [reply]
Thanks for the answer - it wasn't clear perhaps - but I simply want to increse the volume.

I have some sample Arduino code that works. It simulates pulsing the buzzer. How do I do the same in micropython - using pwm?
```

...
pinMode(20, OUTPUT); // buzzer inverted
pinMode(21, OUTPUT); // buzzer

...

  if (buzzer == LOW) // compare to low, not setting to low
  {
    for (int i = 0; i <= 400; i++) // at 4kHz 400 count = 100ms, approx time of shortest button press
    {
      digitalWrite(21, HIGH);  // sets the pin to 3V3
      digitalWrite(20, LOW);   // sets the pin to 0V
      delayMicroseconds(125);  // wait for 125 microseconds
      digitalWrite(21, LOW);   // sets the pin to 0V, swapping pin state gives 2x voltage across buzzer
      digitalWrite(20, HIGH);  // sets the pin to 3V3
      delayMicroseconds(125);  // wait for 125 microseconds
    }
  }
  else //should not have any dc bias on buzzer
  {
    digitalWrite(21, LOW);  // sets the pin to 0V
    digitalWrite(20, LOW);   // sets the pin to 0V
  }
```

**Reply by adrianblakey** [reply]
After a dfew searches - I think I need to invert the phase of the pwm signal on the second pin by setting the control registers - https://forums.raspberrypi.com/viewtopic.php?t=332255 - is this right?

**Reply by robert-hh** [reply]
There is a invert=True/False keyword option when specifying the PWM signal.

**Reply by 2dof** [reply]
If You want invert polarity then read [The Pico In MicroPython: PWM ](https://www.i-programmer.info/programming/hardware/14751-the-pico-in-micropython-pwm.html?start=1). 

 edit. read about voltage dividets controlled by GPIO ( [resistors-in-voltage-divider-](like https://electronics.stackexchange.com/questions/162251/control-connectivity-of-resistors-in-voltage-divider-arrangement), I think it is better solution than changing polarity in buzzer to increase volume unless You use symmetric power source. )

**Reply by robert-hh** [reply]
It can be made pretty easy just using PWM statements. The code below creates a push-pull signal at pins 2 and 3.
```
from machine import PWM, Pin
p1=PWM(Pin(2), freq=1000, duty_u16=32768)
p2=PWM(Pin(3), freq=1000, duty_u16=32768, invert=True)
```

**Reply by davefes** [reply]
> digitalWrite(21, LOW);   // sets the pin to 0V, swapping pin state gives 2x voltage across buzzer

How does that happen?  Have you still not got only 3V3 across the buzzer?   Maybe, requires a link to the buzzer used.

**Reply by peterhinch** [reply]
@robert-hh That code takes no account of phase. Any delay between setting p1 and p2 will result in a phase shift. At higher frequencies the two pins might end up being in phase with no output at all...

**Reply by GitHubsSilverBullet** [reply]
Pins 20 and 21 are attached to the same PWM slice, thus sharing a common TOP counter. They have no chance not to be in sync if both CC are identical.

**Reply by robert-hh** [reply]
@peterhinch  I entered the sample code manually. And the output was perfectly in sync, since both pins used the same slice, as @GitHubsSilverBullet  told.

**Reply by peterhinch** [reply, marked as answer]
Apologies. Interesting - I hadn't appreciated that aspect of the RP2.
